Fix subframes disappearing when object is detached/scaled/renamed

felixvd opened the pull request:
### Description

Subframes were being forgotten when objects were detached/removed from the robot, and when they were moved. They also would have overwritten the old subframes during an APPEND call. This PR should fix all of that, and adds a line to the Rviz plugin that says "The object has subframes X, Y, Z."

Subframes are not updated if the object is moved from the Rviz plugin. I am filing another PR about this in a minute.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/)
- [X] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

felixvd commented:
The tests on this PR succeed, only the code coverage is complaining. The tests for the functions affected by this should be implemented in https://github.com/ros-planning/moveit/pull/1757.

v4hn commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1835,12 +1838,32 @@ bool PlanningScene::processCollisionObjectMove(const moveit_msgs::CollisionObjec
+      ROS_WARN_NAMED(LOGNAME, "Object '%s' had subframes, but no new ones were given for Move operation. ",
+                     object.id.c_str());
+    }
+
```
i am not happy with the need to have the names vector transmitted here, but there's no way around that.
at least, it should be a normal use case of this to specify no subframes, even if the world object includes some.

v4hn commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1835,12 +1838,32 @@ bool PlanningScene::processCollisionObjectMove(const moveit_msgs::CollisionObjec
       obj.reset();
       world_->removeObject(object.id);
       world_->addToObject(object.id, shapes, new_poses);
+      world_->setSubframesOfObject(object.id, subframe_poses);
```
isn't updating the subframes logically independent of specifying poses for everything else?

i can't help but think the use case you want to allow here could be much nicer included in APPEND.
maybe we could just add subframes to the warning in the beginning that says they will be ignored?

rhaschke replied:
My first thought was the very same as Michael's here: Why should we couple an update of shapes to the updates of subframes.
On the other hand, moving shapes, typically also invalidates subframes and thus it makes sense to update the latter in the same call. The present code essentially replicates the corresponding code of the `APPEND` mode. A warning is issued when shapes have moved, but no updates (at all) were specified for subframes.

v4hn replied:
> The present code essentially replicates the corresponding code of the APPEND mode.

I would argue the code should sanitize the subframes list and forbid adding new frames here.
After all, if you want to add new frames, you can use `APPEND`.

felixvd replied:
I removed this because the logic will change in #2037

rhaschke commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1822,12 +1807,32 @@ bool PlanningScene::processCollisionObjectMove(const moveit_msgs::CollisionObjec
     collision_detection::World::ObjectConstPtr obj = world_->getObject(object.id);
+
+    // Update subframe poses
+    moveit::core::FixedTransformsMap subframe_poses = obj->subframe_poses_;
```
This whole code snipped on subframe updates is only relevant if a shape update is actually performed below.
Hence, I think this code should be moved into the if condition below and a warning should be issued if a user just tries to move subframes (but no shapes).

v4hn replied:
The data structures actually allow to handle the additional use-case to specify *only* subframe poses and no shapes.
It's not really an important use-case (maybe for tcp online estimation?), but I think we could include it here?
Either way, the current patch need updating as it mixes both interpretations, doing the computations independent and then update them only if new shape poses exist...

rhaschke replied:
If you just want to update subframes, you better use APPEND mode.

felixvd replied:
Yes, the MOVE operation shouldn't add anything to the object. I removed the whole if block and just updated the world object.

felixvd commented:
I split this as requested without any other changes for now.

v4hn commented:
@felixvd is (part of) this still unmerged? could you resolve the conflict?

felixvd commented:
I updated and rebased this. The PR now only fixes the problem that subframes disappear when objects are detached objects from the robot, or scaled/renamed in Rviz. The changes to the MOVE operation will be in #2037 instead.

felixvd commented:
Ping

v4hn commented on moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame_objects.cpp:
```diff
@@ -122,6 +123,7 @@ void MotionPlanningFrame::sceneScaleChanged(int value)
           ps->getWorldNonConst()->addToObject(scaled_object_->id_, shapes::ShapeConstPtr(s),
                                               scaled_object_->shape_poses_[i]);
         }
+        ps->getWorldNonConst()->setSubframesOfObject(scaled_object_->id_, subframes);
```
This might be a matter of personal preference, but shouldn't the subframes be scaled as well?
I guess there is no reasonable way to do that until your main pose request is merged...
Could you add a TODO in the code here though?

felixvd replied:
Good idea

v4hn commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -120,6 +120,12 @@ class AttachedBody
     return subframe_poses_;
   }
 
+  /** \brief Get subframes of this object (relative to the link) */
```
This is probably just another issue with your ongoing pose work, but why are *global* frames documented as *relative to the link*? Maybe extend/change the comment?

felixvd replied:
I think I just forgot to update this comment, my bad.

v4hn commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1593,51 +1593,35 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
-      }
+      const moveit::core::AttachedBody* body = robot_state_->getAttachedBody(object.object.id);
+      if (body)
+        attached_bodies.push_back(body);
```
Could you please check for whether the id is actually attached to `object.link_name` and trigger a warning if that's not the case?
The previous version would simply ignore the object if a wrong link is specified, but a warning is of course preferred. :)

felixvd replied:
Sure, seems like a rare error though. Please have a look and confirm that this is what you meant.

v4hn replied:
The condition is good, but I would say we either treat it as a proper error the user has to fix before they can continue (so the method should return), or we specify in the warning that the object attached to the other link is used instead.

felixvd replied:
I don't know which way will be better, but I chose to make it an error and fixed the condition (it would have fired for the default value otherwise).

v4hn commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1773,7 +1757,7 @@ bool PlanningScene::processCollisionObjectAdd(const moveit_msgs::CollisionObject
     setObjectType(object.id, object.type);
 
   // Add subframes
```
```suggestion
  // Add subframes to the newly created or (possibly) modified object
```
It took me a moment to understand why this `world_->getObject(id)` call always returns a valid object.

felixvd replied:
Done. Yes, the world object structure is a little awkward.

v4hn commented:
> Perfect

Not really. I just had to fix one misuse of `ROS_ERROR_NAMED`.

The second review might take some time because Robert is on vacation.
That is to say, unless some other maintainer is willing to look at this. :-)

felixvd commented:
I see. I figured it would also concatenate. The more you know

tylerjw commented on moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame_objects.cpp:
```diff
@@ -159,6 +159,7 @@ void MotionPlanningFrame::sceneScaleChanged(int value)
     {
       if (ps->getWorld()->hasObject(scaled_object_->id_))
       {
+        const moveit::core::FixedTransformsMap subframes = scaled_object_->subframe_poses_;  // Keep subframes
```
[nit] Do you need to copy this to a local if it is const?  Could this be a `const&`?

Or maybe it shouldn't be const if you are planning on modifying it when you scale it?

tylerjw replied:
:+1:
